contractcourt/anchor_resolver.go

```diff
@@ -0,0 +1,159 @@
+	resolved bool
+
+	// broadcastHeight is the height that the original contract was
+	// broadcast to the main-chain at. We'll use this value to bound any
+	// historical queries to the chain for spends/confirmations.
+	broadcastHeight uint32
+
+	// chanPoint is the channel point of the original contract.
+	chanPoint wire.OutPoint
+
+	contractResolverKit
+}
+
+// newAnchorResolver instantiates a new anchor resolver.
+func newAnchorResolver(anchorSignDescriptor input.SignDescriptor,
+	anchor wire.OutPoint, broadcastHeight uint32,
+	chanPoint wire.OutPoint, resCfg ResolverConfig) *anchorResolver {
+
+	r := &anchorResolver{
+		contractResolverKit:  *newContractResolverKit(resCfg),
+		anchorSignDescriptor: anchorSignDescriptor,
+		anchor:               anchor,
+		broadcastHeight:      broadcastHeight,
+		chanPoint:            chanPoint,
+	}
+
+	r.initLogger(r)
+
+	return r
+}
+
+// ResolverKey returns an identifier which should be globally unique for this
+// particular resolver within the chain the original contract resides within.
+func (c *anchorResolver) ResolverKey() []byte {
+	// The anchor resolver is stateless and doesn't need a database key.
+
+	return nil
+}
+
+// Resolve offers the anchor output to the sweeper and waits for it to be swept.
+//
+// NOTE: This function MUST be run as a goroutine.
+func (c *anchorResolver) Resolve() (ContractResolver, error) {
+	// Attempt to update the sweep parameters to the post-confirmation
+	// situation. We don't want to force sweep anymore, because the anchor
+	// lost its special purpose to get the commitment confirmed. It is just
+	// an output that we want to sweep only if it is economical to do so. An
+	// exclusive group is not necessary either, because we know that this is
+	// the only anchor that can be swept.
+	relayFeeRate := c.Sweeper.RelayFeePerKW()
+
+	postConfSweepParams := sweep.Params{
+		Fee: sweep.FeePreference{
+			FeeRate: relayFeeRate,
+		},
+	}
+
+	resultChan, err := c.Sweeper.UpdateParams(
+		c.anchor, postConfSweepParams,
+	)
+
+	// After a restart or when the remote force closes, the sweeper is not
+	// yet aware of the anchor. In that case, offer it as a new input to the
+	// sweeper.
+	if err == lnwallet.ErrNotMine {
+		anchorInput := input.MakeBaseInput(
+			&c.anchor,
+			input.CommitmentAnchor,
+			&c.anchorSignDescriptor,
+			c.broadcastHeight,
+		)
+
+		resultChan, err = c.Sweeper.SweepInput(
+			&anchorInput, postConfSweepParams,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	select {
+	case sweepRes := <-resultChan:
+		switch sweepRes.Err {
+
+		// Anchor was swept successfully.
+		case nil:
+			c.log.Debugf("anchor swept by tx %v",
+				sweepRes.Tx.TxHash())
+
+		// Anchor was swept by someone else. This is possible after the
+		// 16 block csv lock.
+		case sweep.ErrRemoteSpend:
+			c.log.Warnf("our anchor spent by someone else")
+
+		// The sweeper gave up on sweeping the anchor. We consider the
+		// anchor as being lost.
+		case sweep.ErrTooManyAttempts:
```

Comment:
when is this being triggered?

Reply:
comment added

Reply:
Ok for now, but in the future we probably want to stop trying exactly when 16 blocks deep, and start attempting  a no sig sweep instead.